Only seed UI test scripts on adhoc

uponthesun opened the pull request:
With this change, when environment is `adhoc`, `seed:all` will perform UI test seeding instead of full seeding.

Reduces adhoc startup time from 80 minutes to 50 minutes - see stack `reduce-adhoc-seed2`.

Also includes changes to output timing information on subtasks for seeding, to make debugging easier if seeding time / adhoc startup time becomes a problem again in the future. If this is a bad idea, I can remove this part.

This implementation doesn't seem like a clean / "rake-y" way to do it, in particular moving from using rake's dependency system to imperative programming, but I couldn't find an easy way to add timing info on seed times without it. Let me know if you think this tradeoff isn't worth it / there's a better way to do it and I can change it back.

davidsbailey commented on dashboard/lib/tasks/seed.rake:
```diff
@@ -148,15 +147,16 @@ namespace :seed do
 
-  task scripts: SCRIPTS_DEPENDENCIES do
+  task scripts: :environment do
+    time_subtasks(SCRIPTS_DEPENDENCIES)
```
It looks like there is more going on here than what's listed in the PR description. why are we timing this? why only for this rake target and not the next two?

uponthesun replied:
Updated PR description.

And good point on the inconsistency - if you're on board with this change overall I'll apply it to the next two also (or I'll get rid of it for all three).

davidsbailey commented on dashboard/lib/tasks/seed.rake:
```diff
@@ -148,15 +147,16 @@ namespace :seed do
   end
 
-  task scripts_incremental: SCRIPTS_DEPENDENCIES do
+  task scripts_incremental: SCRIPTS_DEPENDENCIES + [:environment] do
```
I may be misreading, but it doesn't look necessary to explicitly include `[:environment]` here and below. worth trying without?

uponthesun replied:
I added it because it was previously included in `SCRIPT_DEPENDENCIES` - does it look to you like it shouldn't be needed? If so I can try without.

davidsbailey replied:
I mentioned it because it looks like many subtasks already depend on it. more generally, why is it important to exclude from the timing measurements?

uponthesun commented:
> I think I am missing the "why" here -- typically we do not commit performance debugging code. can we accomplish the same result using miniprofiler? https://github.com/code-dot-org/code-dot-org/blob/2dc83f33fc48609eecfed9cf1cfb3d8c4753b2f8/dashboard/config/initializers/mini_profiler.rb

I think of this less as "performance debugging code" or a profiler and more like making the logging output more useful. The rake tasks don't add timestamps or timing information by default and so it's difficult to see which subtask is taking a long time, or to see when regressions happen. Here are the lines of output added:
```
              Starting seed:ui_test at 2020-03-02 20:52:32 +0000
              Starting seed:videos at 2020-03-02 20:52:32 +0000
              Finished seed:videos at 2020-03-02 20:52:46 +0000, took 13.689521927 seconds
              Starting seed:concepts at 2020-03-02 20:52:46 +0000
              Finished seed:concepts at 2020-03-02 20:52:46 +0000, took 0.017142046 seconds
              Starting seed:scripts_ui_tests at 2020-03-02 20:52:46 +0000
              Finished seed:scripts_ui_tests at 2020-03-02 21:10:12 +0000, took 1046.322821629 seconds
              Starting seed:courses_ui_tests at 2020-03-02 21:10:12 +0000
              Finished seed:courses_ui_tests at 2020-03-02 21:10:12 +0000, took 0.137712073 seconds
              Starting seed:callouts at 2020-03-02 21:10:12 +0000
              Finished seed:callouts at 2020-03-02 21:10:12 +0000, took 0.335636942 seconds
              Starting seed:school_districts at 2020-03-02 21:10:12 +0000
              Finished seed:school_districts at 2020-03-02 21:10:12 +0000, took 0.033001866 seconds
              Starting seed:schools at 2020-03-02 21:10:12 +0000
              Finished seed:schools at 2020-03-02 21:10:13 +0000, took 0.052728309 seconds
              Starting seed:secret_words at 2020-03-02 21:10:13 +0000** Invoke seed:secret_words (first_time)
              Finished seed:secret_words at 2020-03-02 21:10:13 +0000, took 0.022389288 seconds
              Starting seed:secret_pictures at 2020-03-02 21:10:13 +0000
              Finished seed:secret_pictures at 2020-03-02 21:10:13 +0000, took 0.025564655 seconds
              Starting seed:donors at 2020-03-02 21:10:13 +0000
              Finished seed:donors at 2020-03-02 21:10:13 +0000, took 0.058212677 seconds
              Starting seed:donor_schools at 2020-03-02 21:10:13 +0000
              Finished seed:donor_schools at 2020-03-02 21:10:13 +0000, took 0.254839908 seconds
              Finished seed:ui_test at 2020-03-02 21:10:13 +0000, took 1060.949959041 seconds
```

These are visible in adhoc startup logs at `/var/log/chef-bootstrap-debug.log`, in drone output, etc.

That said I'm open to reasons why this shouldn't be added.

davidsbailey commented:
> These are visible in adhoc startup logs at `/var/log/chef-bootstrap-debug.log`, in drone output, etc.

Will these show up in `/var/log/chef-bootstrap.log` ? asking because those logs appear to have timestamps already in the one adhoc I looked at.

Will the timing info start showing up for various environments in slack too?

I'm happy to provide input here but would also like to hear what @wjordan thinks.

uponthesun commented:
> Will these show up in `/var/log/chef-bootstrap.log` ? asking because those logs appear to have timestamps already in the one adhoc I looked at.

No - the output in `chef-bootstrap.log` is pretty coarse and doesn't have the output from our application, it just has chef-specific stuff.

> Will the timing info start showing up for various environments in slack too?

No, this just goes to standard out. Not sure how useful it would be for managed environments since those aren't seeding from scratch, and this output is pretty verbose compared to what we have in slack right now.

wjordan commented:
I think the general idea of adding timing-logging info to certain long-running rake tasks is good and helpful. For the implementation however, I'm not sure I like the changes needed to the rake tasks to manually invoke the dependencies within the block rather than through the normal Rake syntax for dependencies.

Could we spend a bit of time looking into whether it's possible to add an extra method (`timed_task`), or support some extra option (`timed: true`) to make the interface a bit cleaner?

If it's not easily done I think the current implementation is fine, just wondering if something simpler is possible.

The change to only seed UI test scripts on adhoc is also :+1: - though let's not forget to give a heads up to the broader team about this change since it changes adhoc behavior quite a bit (should be better in the general case).

We can also consider some switch to use the original behavior (in case someone needs an adhoc with all scripts seeded?), but I think I'd prefer to leave that until we actually get such a request (which may never come).

uponthesun commented:
I changed to the implementation suggested by Will. As part of this I also added a new `seed:default` task which is invoked both by `rake build` and `dashboard:setup_db` - the two places as far as I know that we call `seed:all` from. The new `seed:default` task decides whether to do a full seed or UI test seed based on the environment. This allows a user to still run `seed:all` on an adhoc if they actually need the full seed.

I also removed the `cache_ui_test_data` since it doesn't seem to do anything.

davidsbailey commented on dashboard/lib/tasks/seed.rake:
```diff
@@ -333,15 +354,19 @@ namespace :seed do
-  task :cache_ui_test_data do
-  end
+  FULL_SEED_TASKS = [:videos, :concepts, :scripts, :callouts, :school_districts, :schools, :secret_words, :secret_pictures, :courses, :ap_school_codes, :ap_cs_offerings, :ib_school_codes, :ib_cs_offerings, :state_cs_offerings, :donors, :donor_schools, :foorms].freeze
+  UI_TEST_SEED_TASKS = [:videos, :concepts, :scripts_ui_tests, :courses_ui_tests, :callouts, :school_districts, :schools, :secret_words, :secret_pictures, :donors, :donor_schools].freeze
```
I found these a bit hard to compare visually, because the order is different. If it's not too annoying, could `:courses` be moved to just after `:scripts` in the first list so that the orders match?

uponthesun replied:
I think I copy / pasted these directly, I'm ok with making an improvement though. An even better improvement might be to refactor the shared tasks into a constant, and then build `FULL_SEED_TASKS` and `UI_TEST_SEED_TASKS` by concatenating the ones that differ. I'll try that out.

uponthesun replied:
Never mind, it gets too complicated with `seed:incremental` in the picture too. I assume order also matters so I don't want to mess with that too much. I'll just make the change ou suggested.

davidsbailey commented:
Thank you @uponthesun , this definitely improves readability, for me anyway :)
